### PR TITLE
mimir : calculate city.weight and use to sort elastic search result

### PR DESCRIPTION
**Pull request by kadhikari**

### CODE REVIEW

File: libs/mimir/src/objects.rs.in [hunk 1]
```diff
@@ -66,32 +69,42 @@ impl<'a, T: EsId> EsId for &'a T {
...
+
 #[derive(Serialize, Deserialize, Debug, Clone)]
 pub struct Admin {
     pub id: String,
     pub insee: String,
     pub level: u32,
     pub name: String,
     pub zip_code: String,
-    pub weight: u32,
+    #[serde(serialize_with="custom_cell_serialize", skip_deserializing)]
```
Comment (antoine-de) [inline libs/mimir/src/objects.rs.in]: can you add a comment about why we don't need to deserialize it ?

File: src/bin/osm2mimir.rs [hunk 2]
```diff
@@ -295,17 +297,37 @@ fn main() {
...
     let admins = administrative_regions(&mut parsed_pbf, levels);
-    let nb_admins = rubber.index("admin", &args.flag_dataset, admins.iter())
-                          .unwrap();
-    info!("Nb of indexed admin: {}", nb_admins);
 
+    let mut streets = streets(&mut parsed_pbf, admins.clone());
+    
+    for st in &mut streets {
+    	for admin in &mut st.administrative_regions {
+    		if admin.level == 8 {
```
Comment (antoine-de) [inline src/bin/osm2mimir.rs]: we don't want to have the level `8`  hardcoded in the code.

File: libs/mimir/src/objects.rs.in (hunk 1 above)
Comment (kadhikari) [inline libs/mimir/src/objects.rs.in]: ok

File: libs/bragi/src/query.rs [hunk 3]
```diff
@@ -100,6 +102,7 @@ fn query(q: &String, cnx: &String, match_type: &str) -> Result<Vec<mimir::Place>
     let result: SearchResult<serde_json::Value> = try!(client.search_query()
                                                        .with_indexes(&["munin"])
                                                        .with_query(&final_query)
+                                                       .with_sort(&Sort::field_order("weight", Order::Desc))
```
Comment (TeXitoi) [inline libs/bragi/src/query.rs]: That's too strong. We should add something in the request, but sorting only on that will break a lot of thing as ngram matching.

Comment (TeXitoi) [inline libs/bragi/src/query.rs]: We can take inspiration of what was done in munin, It already has this kind of feature (but counting the number of house number of the street)

File: libs/mimir/src/objects.rs.in [hunk 4]
```diff
@@ -66,32 +69,44 @@ impl<'a, T: EsId> EsId for &'a T {
...
+    }
+}
 
 #[derive(Serialize, Deserialize, Debug, Clone)]
 pub struct Coord {
     pub lat: f64,
     pub lon: f64,
 }
 
+
```
Comment (TeXitoi) [inline libs/mimir/src/objects.rs.in]: useless new line.

File: src/bin/osm2mimir.rs [hunk 5]
```diff
@@ -289,23 +296,44 @@ fn main() {
...
     debug!("creation of indexes");
     let mut rubber = Rubber::new(&args.flag_connection_string);
 
     debug!("importing adminstrative region into Mimir");
     let admins = administrative_regions(&mut parsed_pbf, levels);
-    let nb_admins = rubber.index("admin", &args.flag_dataset, admins.iter())
-                          .unwrap();
-    info!("Nb of indexed admin: {}", nb_admins);
 
+    let mut streets = streets(&mut parsed_pbf, admins.clone());
```
Comment (TeXitoi) [inline src/bin/osm2mimir.rs]: You must not clone here. The solution is that `fn streets` and `fn make_admin_geofinder` take a &Vec

**Comment by kadhikari** [comment]
Ticket concerned : http://jira.canaltp.fr/browse/NAVP-594
Note: A new ticket http://jira.canaltp.fr/browse/NAVP-617 is created for tuning elastic search.

File: libs/mimir/src/objects.rs.in [hunk 6]
```diff
@@ -80,18 +93,19 @@ pub struct Admin {
...
-    fn id(&self) -> &str {
-        &self.id
-    }
-    fn incr(&mut self) {
-        self.weight += 1;
-    }
+
+fn custom_cell_serialize<S>(cell: &Cell<u32>, serializer: &mut S) -> Result<(), S::Error> where S: Serializer {
+	// we can serialize the cell as a u32, since the reference is important only while loading the data
+	// and not in ES
```
Comment (TeXitoi) [inline libs/mimir/src/objects.rs.in]: and in ES but not in bragi

Comment (kadhikari) [inline libs/mimir/src/objects.rs.in]: ok

File: libs/mimir/src/objects.rs.in [hunk 7]
```diff
@@ -105,7 +104,7 @@ pub struct Admin {
 
 fn custom_cell_serialize<S>(cell: &Cell<u32>, serializer: &mut S) -> Result<(), S::Error> where S: Serializer {
 	// we can serialize the cell as a u32, since the reference is important only while loading the data
-	// and not in ES
+	// and not in ES but not in bragi
```
Comment (TeXitoi) [inline libs/mimir/src/objects.rs.in]: weight are used in ES, so "and ~~not~~ in ES" ;-)

**Comment by TeXitoi** [comment]
OK for me with the comment corrected

File: libs/mimir/src/objects.rs.in (hunk 7 above)
Comment (kadhikari) [inline libs/mimir/src/objects.rs.in]: my mistake

**Pull request merged**